### PR TITLE
Add Snowflake and BigQuery warehouse query connectors

### LINKED ISSUE #205: Add Snowflake / BigQuery data warehouse connector
## Summary

Add connector(s) for data warehouse platforms (Snowflake and/or BigQuery). Typically read-only, high enterprise value for agents in analytics and reporting workflows.

## Actions

- [x] `snowflake.query` — Execute parameterized read-only SQL queries
- [x] `bigquery.query` — Execute parameterized read-only SQL queries

## Design Considerations

- **Read-only by default** — data warehouses are primarily for querying, not writing
- **Cost controls** — cap query scan size or compute credits per execution
- **Result size limits** — warehouse queries can return massive result sets
- **Statement timeouts** — prevent long-running analytical queries from blocking
- **Credentials** — Snowflake (user/password or key pair), BigQuery (service account JSON) stored in vault

## Notes

- May make sense as two separate connectors given different auth models
- High enterprise value — analytics teams using agents for reporting
- Consider starting with Snowflake (simpler auth) then adding BigQuery

### PR DESCRIPTION
**Pull request by chiedocursor**
## Summary

- Add `snowflake.query` for read-only SELECT/WITH queries via the Snowflake Go driver: DSN in `connection_string`, optional `private_key_pem` (RSA key set on `gosnowflake.Config.PrivateKey` and opened with `sql.OpenDB(NewConnector)` — not embedded in the DSN string), `ALTER SESSION` statement timeout, read-only transaction, and `max_rows` cap.
- Add `bigquery.query` using `service_account_json`: `?` placeholders (outside strings/comments) map to `@p0`, `@p1`, …; `validateReadOnlyBigQuerySQL` rejects `WITH …` DML, multi-statement `;`, and dangerous top-level keywords; `max_bytes_billed` (default 10 GiB), job timeout, optional `default_dataset` / `query_job_location`, row cap.
- Dependencies: `cloud.google.com/go/bigquery` **v1.74.0**, `github.com/snowflakedb/gosnowflake` **v1.10.0**, module `go` directive **1.25.8** (matches repo toolchain). Merged `main` (includes Firestore connector); built-in connector registry test expects **55** connectors.

Closes #205

### Developer
- [x] `go test ./connectors/snowflake/... ./connectors/bigquery/...`
- [x] `go test ./connectors/...` (includes registry count)
- [x] `make build` (frontend + mobile generate + Go binary)

### Manual testing
- [ ] Enable `snowflake` / `bigquery` for an agent in staging, store real vault credentials, run a small parameterized SELECT and confirm row cap + timeout behavior.
- [ ] BigQuery: confirm `max_bytes_billed` rejects expensive queries as expected; Snowflake: confirm key-pair auth end-to-end.

https://claude.ai/code

<div><a href="https://cursor.com/agents/bc-6ee26b77-64a7-4a58-95df-5c9c225cc459"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-6ee26b77-64a7-4a58-95df-5c9c225cc459"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

### CODE REVIEW

**Review by chiedo — APPROVED** [review]

File: connectors/bigquery/query.go [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+}
+
+func (p *queryParams) validate() error {
+	if strings.TrimSpace(p.SQL) == "" {
+		return &connectors.ValidationError{Message: "missing required parameter: sql"}
+	}
+	normalized := strings.TrimSpace(strings.ToUpper(p.SQL))
+	if !strings.HasPrefix(normalized, "SELECT") && !strings.HasPrefix(normalized, "WITH") {
+		return &connectors.ValidationError{Message: "only SELECT queries are allowed (query must start with SELECT or WITH)"}
+	}
```
[the remark this answers is not in the document]
Comment (chiedocursor) [reply connectors/bigquery/query.go]: **Updates pushed (commit 3a4f2fc1):**

- **BigQuery DML / WITH:** Added `validateReadOnlyBigQuerySQL` — parses `WITH` CTEs, requires main statement `SELECT`, rejects `WITH … DELETE/INSERT/UPDATE/MERGE`, blocks top-level dangerous keywords outside parens, and rejects `;` multi-statement.
- **Placeholders:** `?` is only counted/replaced outside strings/comments (state machine in `sql_scan.go`).
- **go.mod:** `go 1.25.8` + `cloud.google.com/go/bigquery v1.74.0`.
- **Snowflake:** `sql.OpenDB(gosnowflake.NewConnector(...))` with `Config.PrivateKey` from vault PEM (no key in DSN string).

Also fixed `connectors/square` truncate test to match `TruncateUTF8` suffix.

File: go.mod [hunk 2]
```diff
@@ -3,6 +3,7 @@ module github.com/supersuit-tech/permission-slip-web
 go 1.25.8
 
 require (
```
[the remark this answers is not in the document]
Comment (chiedocursor) [reply go.mod:5]: Updated the PR description to match the current diff: BigQuery **v1.74.0**, **go 1.25.8**, and the stricter read-only / Snowflake connector notes.

File: connectors/bigquery/sql_scan.go [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+			j++
+		}
+		s = strings.TrimLeftFunc(s[j:], unicode.IsSpace)
+		if !strings.HasPrefix(s, "AS") {
+			return "", &connectors.ValidationError{Message: "invalid WITH clause: expected AS"}
+		}
+		if len(s) > 2 && isIdentCont(s[2]) {
+			return "", &connectors.ValidationError{Message: "invalid WITH clause: expected AS keyword"}
+		}
+		s = strings.TrimLeftFunc(s[3:], unicode.IsSpace)
```
[the remark this answers is not in the document]
Comment (chiedocursor) [reply connectors/bigquery/sql_scan.go]: Fixed: advance past `AS` with `s[2:]` then trim (no panic on `AS` alone; `AS(` works). Added `WITH` / `SELECT` word-boundary checks.

File: connectors/bigquery/sql_scan.go [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+	s := strings.TrimSpace(scrubbed)
+	if s == "" {
+		return &connectors.ValidationError{Message: "missing required parameter: sql"}
+	}
+	u := strings.ToUpper(s)
+	if err := rejectMultiStatement(u); err != nil {
+		return err
+	}
+	if strings.HasPrefix(u, "WITH") {
+		rest := strings.TrimSpace(u[len("WITH"):])
```
[the remark this answers is not in the document]
Comment (chiedocursor) [reply connectors/bigquery/sql_scan.go]: Addressed in the same commit: `WITH` branch only when the 5th char is not `isIdentCont`, and `SELECT` similarly at position 6.

File: connectors/snowflake/query.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	if !strings.HasPrefix(normalized, "SELECT") && !strings.HasPrefix(normalized, "WITH") {
+		return &connectors.ValidationError{Message: "only SELECT queries are allowed (query must start with SELECT or WITH)"}
+	}
+	if p.MaxRows < 0 {
+		return &connectors.ValidationError{Message: "max_rows must be positive"}
+	}
+	if p.MaxRows > 10000 {
+		return &connectors.ValidationError{Message: "max_rows cannot exceed 10000"}
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (chiedocursor) [reply connectors/snowflake/query.go:36]: Snowflake now uses `sqldb.ValidateReadOnlyWarehouseSQL` (shared with BigQuery) so `WITH …` DML fails with ValidationError before hitting the DB.

File: connectors/snowflake/query.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+		return nil, mapSnowflakeError(err, "setting statement timeout")
+	}
+
+	tx, err := db.BeginTx(ctx, &sql.TxOptions{ReadOnly: true})
+	if err != nil {
+		return nil, mapSnowflakeError(err, "starting read-only transaction")
+	}
+	defer tx.Rollback() //nolint:errcheck
+
+	rows, err := tx.QueryContext(ctx, params.SQL, params.Params...)
```
[the remark this answers is not in the document]
Comment (chiedocursor) [reply connectors/snowflake/query.go]: Params are coerced via `sqldb.CoerceJSONParamValue` (same logic BigQuery used) before `QueryContext`.

**Comment by chiedocursor** [comment]
## 🔧 CI / audit remediation

### Remediation round 1

- **audit.yml** — conclusion `failure` — govulncheck flagged gosnowflake@v1.10.0 (GO-2025-3650) and jose2go@v1.6.0 (GO-2025-4123); upgraded to gosnowflake v1.13.3 and jose2go v1.7.0 (`b81f226d`)

File: connectors/bigquery/sql_scan.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+				i++
+			}
+		case c == '/' && i+1 < len(sql) && sql[i+1] == '*':
+			end := strings.Index(sql[i+2:], "*/")
+			if end < 0 {
+				b.WriteString(sql[i:])
+				return b.String(), nil
+			}
+			end += i + 2
+			b.WriteString(sql[i : end+2])
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/bigquery/sql_scan.go:44]: Fixed in PR #849. Added triple-quote detection (`"""..."""` and `'''...'''`) to both `findBarePlaceholders` in `sql_scan.go` and `scrubSQLMasking` in `readonly_sql.go`. The triple-quote case is checked before the single-quote/double-quote cases so it takes priority. Tests added for all three scenarios (triple-single, triple-double, and triple-with-bare-placeholder).

File: pkg/sqldb/json_params.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+// CoerceJSONParamValue normalizes values produced by encoding/json.Unmarshal
+// (e.g. whole-number float64 → int64) before passing them to database/sql.
+func CoerceJSONParamValue(v interface{}) interface{} {
+	switch x := v.(type) {
+	case float64:
+		if !math.IsNaN(x) && !math.IsInf(x, 0) && x == math.Trunc(x) && x >= math.MinInt64 && x <= math.MaxInt64 {
+			return int64(x)
+		}
+		return x
+	case json.Number:
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply pkg/sqldb/json_params.go:17]: Fixed in PR #849. Changed the upper-bound comparison to `x < float64(math.MaxInt64)+1` (which is exactly `9223372036854775808.0`) and explicitly cast the lower bound to `float64(math.MinInt64)`. Values at the boundary now correctly remain as `float64` instead of silently wrapping to `MinInt64`. Test added for the boundary value.

File: connectors/bigquery/sql_scan.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+			}
+		case c == '?':
+			n++
+			i++
+		default:
+			i++
+		}
+	}
+	return n
+}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/bigquery/sql_scan.go:180]: Fixed in PR #849. Extracted a shared `findBarePlaceholders(sql string) []int` function that contains the single scanning state machine. Both `countParamPlaceholders` and `translatePlaceholders` now delegate to it — `countParamPlaceholders` returns `len(positions)` and `translatePlaceholders` splices the string at the returned offsets. This eliminated ~70 lines of duplicated code and ensures triple-quote support (and any future fix) only needs to be implemented once.

File: pkg/sqldb/readonly_sql.go [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+					b.WriteByte(' ')
+				}
+				i += w
+			}
+		case c == '`':
+			b.WriteByte(' ')
+			i++
+			for i < len(sql) && sql[i] != '`' {
+				b.WriteByte(' ')
+				i++
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply pkg/sqldb/readonly_sql.go:135]: Fixed in PR #849. Changed `scrubSQLMasking` to replace backtick-quoted content with underscores (`_`) instead of spaces. This preserves a valid identifier token in the scrubbed output, so `mainKeywordAfterCTEs` correctly parses `` WITH `my-dataset` AS (...) `` as having a CTE name (the underscored placeholder) followed by `AS`. Test added for the backtick CTE name case.

File: pkg/sqldb/readonly_sql.go [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+			for i+1 < len(sql) {
+				if sql[i] == '*' && sql[i+1] == '/' {
+					b.WriteByte(' ')
+					b.WriteByte(' ')
+					i += 2
+					break
+				}
+				b.WriteByte(' ')
+				i++
+			}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply pkg/sqldb/readonly_sql.go:75]: Fixed in PR #849. Changed the inner loop from `for i+1 < len(sql)` to `for i < len(sql)` with the terminator check `if i+1 < len(sql) && ...` moved inside the loop body. This ensures all bytes of an unterminated block comment are masked, including the final byte. Test added to verify no comment bytes leak through.
